File: day9/py/main.py
from time import perf_counter_ns

def getInput(lines):
    nums = []
    for line in lines:
        nums.append([int(x) for x in line.split(' ')])
    return nums

def extra_end(triangle):
    return sum (x[-1] for x in triangle)

def extra_begin(triangle):
    flip = True
    res = 0
    for n in triangle:
        if flip:
            res += n[0]
        else:
            res -= n[0]
        flip ^= True
    return res

def oasis(nums, extra):
    triangle = []
    triangle.append(nums)
    while (True):
        temp = []
        for i in range(1, len(triangle[-1])):
            temp.append(triangle[-1][i] - triangle[-1][i - 1])
        triangle.append(temp)
        if all(x == 0 for x in temp): break
    return extra(triangle)

def part1(input):
    return sum([oasis(x, extra_end) for x in input])

def part2(input):
    return sum([oasis(x, extra_begin) for x in input])

def measure(func, arg, msg="", unit="us"):
    match unit:
        case "ns": scale = 1
        case "us": scale = 1_000
        case "ms": scale = 1_000_000
        case "s": scale = 1_000_000_000
        case _: return 0
    start = perf_counter_ns()
    res = func(arg)
    end = ((perf_counter_ns() - start) / scale)
    if msg != "": print(msg, res)
    print(f"time[{unit}] = {end:.2f}")

input = getInput(open("../input.txt", "r").readlines())
print("part1 = ", part1(input))
print("part2 = ", part2(input))

File: day9/cs/main.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace aoc
{
    internal class main
    {
        delegate int Extra(List<List<int>> triangle);

        static int extraEnd(List<List<int>> triangle)
        {
            int sum = 0;
            foreach (List<int> nums in triangle)
            {
                sum += nums[nums.Count - 1];
            }
            return sum;
        }

        static int extraBegin(List<List<int>> triangle)
        {
            bool flip = true;
            int sum = 0;
            foreach (List<int> nums in triangle)
            {
                if (flip) sum += nums[0];
                else sum -= nums[0];
                flip = !flip;
            }
            return sum;
        }

        static List<List<int>> parseInput(string fileName)
        {
            List<List<int>> input = new List<List<int>>();
            StreamReader sr = new StreamReader(fileName);
            string line;
            while ((line = sr.ReadLine()) != null)
            {
                List<int> nums = new List<int>();
                foreach (string strnum in line.Split(' '))
                { 
                    nums.Add(int.Parse(strnum));
                }
                input.Add(nums);
            }
            return input;
        }
        
        static int solution(List<int> nums, Extra extra)
        {
            List<List<int>> triangle = new List<List<int>>();
            triangle.Add(nums);
            while (true)
            {
                List<int> subs = new List<int>();
                for (int i = 1, last = triangle.Count - 1; i < triangle[last].Count; i++)
                {
                    subs.Add(triangle[last][i] - triangle[last][i - 1]);
                }
                triangle.Add(subs);
                if (subs.All(x => x == 0)) { break; }
            }
            return extra(triangle);
        }

        static int part1(List<List<int>> input)
        {
            int sum = 0;
            foreach(List<int> nums in input)
            {
                sum += solution(nums, extraEnd);
            }
            return sum;
        }

        static int part2(List<List<int>> input)
        {
            int sum = 0;
            foreach(List<int> nums in input)
            {
                sum += solution(nums, extraBegin);
            }
            return sum;
        }
        static void show<T1, T2>(string s, Func<T1, T2> func, T1 arg)
        {
            Stopwatch timer = Stopwatch.StartNew();
            T2 res = func(arg);
            timer.Stop();
            Console.WriteLine(String.Format(s + "{0}", res));
            Console.WriteLine(String.Format("time[ms] = {0}", timer.Elapsed.TotalMilliseconds));
        }

        static void Main(string[] args)
        {
            List<List<int>> input = parseInput(@"..\..\..\input.txt");
            Console.WriteLine("part1 = {0}", part1(input));
            Console.WriteLine("part2 = {0}", part2(input));
        }
    }
}
